Add LTI 1.3 instance

trombonekenny opened the pull request:
This PR is the first in the LTI 1.3 support. It introduces the concept and management of an LTI 1.3 instance that is a configuration and trust relationship between another LMS and PL. The pages in this PR are an institutional admin configuration page and some non-authenticated endpoints that support setting up Canvas configurations and JWKS key exchanges.

Uses the institution feature flag `lti13`

trombonekenny commented:
Hmmm, that's weird. I don't remember doing commits in the dev container (that would have tagged 'Dev User'). How do we resolve this?

jonatanschroeder commented:
dec30862a254530d7fd666037d009945d491308c is authored by Dev User. I have rebased it and forced a squash of that commit into the following one.

jonatanschroeder commented:
> [dec3086](https://github.com/PrairieLearn/PrairieLearn/commit/dec30862a254530d7fd666037d009945d491308c) is authored by Dev User. I have rebased it and forced a squash of that commit into the following one.

Didn't work... I'll try something else.

nwalters512 commented on apps/prairielearn/src/ee/pages/institutionAdminLti13/institutionAdminLti13.html.ts:
```diff
@@ -0,0 +1,300 @@
+  `.toString();
+}
+
+const displayLTI13Instance = function (instance, resLocals, platform_defaults) {
```
Missing types. Conventionally, this kind of "component" (borrowing React nomenclature) is `PascalCase` and doesn't include "display" or any kind of verb. You can declare functions directly with `function LTI13Instance(...)` instead of using `const`.

trombonekenny replied:
Addressed in a previous commit.

nwalters512 commented on apps/prairielearn/src/ee/pages/institutionAdminLti13/institutionAdminLti13.ts:
```diff
@@ -0,0 +1,192 @@
+);
+
+router.post(
+  '/:lti13_instance_id?',
```
Conventionally, we prefix route params with `unsafe_` to indicate that they aren't trusted (that is, someone could edit a URL to include an LTI instance ID that's unrelated to the current institution.

trombonekenny commented on apps/prairielearn/src/migrations/20231003211522_lti13_instance__create.sql:
```diff
@@ -0,0 +1,65 @@
+    '{
+   "client_id": "10000000000001",
+   "redirect_uris": [
+      "https://pl.tbkenny.com/pl/lti13_instance/1/auth/callback"
```
Fix me: Don't hard code hosting URLs. Should already be fixed in code and can be removed from here.

nwalters512 commented on apps/prairielearn/src/ee/pages/lti13Instance/lti13Instance.ts:
```diff
@@ -0,0 +1,80 @@
+    ltiConfig.oidc_initiation_url = `${url}/pl/lti13_instance/${res.locals.lti13_instance.id}/auth/login`;
+    ltiConfig.target_link_uri = `${url}/pl/lti13_instance/${res.locals.lti13_instance.id}/auth/callback`;
+
+    ltiConfig.extensions[0].domain = req.get('host') || '';
```
We should probably use the same logic as `getCanonicalHost` so that we use `config.serverCanonicalHost` if it's present. If this expects something without a protocol, we can strip it if needed.

trombonekenny replied:
OK, I'll upload a different take on this.

nwalters512 commented on apps/prairielearn/src/ee/pages/lti13Instance/lti13Instance.ts:
```diff
@@ -0,0 +1,80 @@
+    ltiConfig.public_jwk_url = `${url}/pl/lti13_instance/${res.locals.lti13_instance.id}/jwks`;
+
+    res.setHeader('Content-type', 'application/json');
+    res.end(JSON.stringify(ltiConfig, null, 3));
```
3 spaces?? Let's stick with 2 or 4, ideally 2 to match the API.

nwalters512 commented on apps/prairielearn/src/ee/routers/institutionAdmin.ts:
```diff
@@ -15,17 +19,23 @@ const router = Router({ mergeParams: true });
+    res.locals.urlPrefix = req.baseUrl;
+
+    const hasLti13 = await features.enabled('lti13', { institution_id: req.params.institution_id });
+    res.locals.lti13_enabled = hasLti13 && isEnterprise();
```
No need for an `isEnterprise()` check here - this router should only ever be attached to the server in an enterprise environment, so this is redundant.

nwalters512 commented on apps/prairielearn/src/ee/pages/lti13Instance/lti13Instance.ts:
```diff
@@ -0,0 +1,80 @@
+  asyncHandler(async (req, res) => {
+    const keystore = await jose.JWK.asKeyStore(res.locals.lti13_instance.keystore || []);
+
+    res.setHeader('Content-type', 'application/json; charset=UTF-8');
```
```suggestion
    res.setHeader('Content-Type', 'application/json; charset=UTF-8');
```

nwalters512 commented on apps/prairielearn/src/ee/routers/lti13.ts:
```diff
@@ -0,0 +1,58 @@
+  }),
+);
+
+router.use(logRequest);
```
Is this necessary? The request should already be logged by the same middleware attached to the server as a whole.

trombonekenny replied:
Yes, because the auth routes do their own CSRF, the `pl/lti13_instance` routes come before those middleware lines in server.js.

nwalters512 replied:
I'm wondering if the right thing to do isn't to move `logRequest` earlier into the middleware chain. Tagging in @mwest1066 - is there a good reason that we haven't been logging certain requests?

trombonekenny replied:
Filed issue #8691 so I'll remove it from here.

mwest1066 replied:
Left a comment on that issue. I agree that we can move `logRequest` up to be just after `logResponse`.

nwalters512 reviewed (commented):
This is a first pass; I'll take a look at the functionality once the invalid HTML is corrected and more easily reviewable.

trombonekenny commented on apps/prairielearn/src/ee/pages/institutionAdminLti13/institutionAdminLti13.html.ts:
```diff
@@ -0,0 +1,300 @@
+      Remove LTI 1.3 instance
+    </button>
+  </form>
+</div>
```
Thanks for finding that. I wasn't sure why Prettier was doing it's job.

trombonekenny commented on apps/prairielearn/src/migrations/20231003211522_lti13_instance__create.sql:
```diff
@@ -0,0 +1,65 @@
+    '{
+   "client_id": "10000000000001",
+   "redirect_uris": [
+      "https://pl.tbkenny.com/pl/lti13_instance/1/auth/callback"
```
Nope, that's a bug from previous development. It's not needed and shouldn't have been there.

trombonekenny commented on apps/prairielearn/src/ee/routers/lti13.ts:
```diff
@@ -0,0 +1,58 @@
+const sql = loadSqlEquiv(__filename);
+const router = Router({ mergeParams: true });
+
+router.use(
+  asyncHandler(async (req, res, next) => {
+    // The navbar relies on this property.
+    res.locals.urlPrefix = req.baseUrl;
+    next();
+  }),
+);
```
This branch is a subset of another bigger LTI 1.3 development branch that has more pages (for later PRs). Do you want me to leave this there, or factor it into the future PR (I guess it could also move to their specific subpages.)

nwalters512 replied:
I'd rather remove this for now, yeah.

trombonekenny replied:
Done in a newer commit.

trombonekenny commented on apps/prairielearn/src/ee/pages/institutionAdminLti13/institutionAdminLti13.ts:
```diff
@@ -0,0 +1,192 @@
+      });
+      flash('success', `Platform updated.`);
+
+      // TODO: Saving changes should remove the cached value in auth
```
No, it's for auth, which is the next slated PR.

trombonekenny commented on apps/prairielearn/src/ee/routers/lti13.ts:
```diff
@@ -0,0 +1,58 @@
+      throw new Error(`LTI 1.3 instance ID ${req.params.lti13_instance_id} is unavailable`);
+    }
+
+    res.locals.lti13_instance = lti13Instance;
```
Do you have any example pattern of a middleware that doesn't use `res.locals`? Or you just move the database lookup to each individual page?

SethPoulsen replied:
You put something in the `models` directory that gets the stuff from the database, then you call that from each page.

trombonekenny replied:
Implemented in models in 15be97e, thanks @SethPoulsen

trombonekenny commented on apps/prairielearn/src/migrations/20231003211522_lti13_instance__create.sql:
```diff
@@ -0,0 +1,65 @@
+   "token_endpoint_auth_signing_alg": "RS256"
+}',
+    10
+  );
```
I figured the list of default configs would grow as we supported more platforms and it seemed more flexible to add various ones (or dev/site specific ones like my tbkenny environment) into the database instead of the code base.

This table is only used to populate the options the javascript asset uses to load a different platform default into the `institutionAdminLti13` page.

But I don't have a strong opinion on this either way if you prefer it not be a table.

nwalters512 replied:
Maybe things will shake out differently in practice, but I can't see why we'd want to add custom defaults in one production environment and not all the rest.

If you really want this to be configurable for local dev, consider defining extra ones in the config file. In production, if we did have environment-specific things, we'd want them in configs anyways so that they can be version-controlled (unlike one-off manually-run database migrations).

mwest1066 replied:
I agree we don't want this table in the DB.

trombonekenny replied:
Moved to config in 875a68a2a

trombonekenny commented on apps/prairielearn/src/ee/pages/lti13Instance/lti13Instance.ts:
```diff
@@ -0,0 +1,80 @@
+
+    const url = getCanonicalHost(req);
+
+    ltiConfig.oidc_initiation_url = `${url}/pl/lti13_instance/${res.locals.lti13_instance.id}/auth/login`;
```
I agree with the mutation comment, good catch.

I don't see anywhere where we're using `structuredClone` in PL or PT. I get a `ReferenceError: structuredClone is not defined` when I try to use it. What do I need to load first to source it from?

nwalters512 replied:
Looks like it isn't present in the version of Node we're using 😢. [`_.cloneDeep`](https://lodash.com/docs/4.17.15#cloneDeep) is probably the way to go for now.

trombonekenny replied:
Switched to using cloneDeep

trombonekenny commented on apps/prairielearn/src/ee/routers/lti13.ts:
```diff
@@ -0,0 +1,58 @@
+import { Router } from 'express';
+import lti13InstanceRouter from '../pages/lti13Instance/lti13Instance';
```
Changed to `lti13InstancePages`

trombonekenny commented on apps/prairielearn/src/ee/routers/lti13.ts:
```diff
@@ -0,0 +1,58 @@
+
+router.use(logRequest);
+
+//
+// Middleware to load the lti13_instance and error check
+//
```
The other pages I'll add under this (auth and the course selection pages) will use it too. lti13InstanceRouter is misleading as this is the router for the LTI 1.3 pages so I renamed it to `lti13InstancePages` (the only current endpoint, and only used for getting data out of the instance for configuration).

trombonekenny commented on apps/prairielearn/src/ee/routers/lti13.ts:
```diff
@@ -0,0 +1,58 @@
+  }),
+);
+
+// router is a catch all for some small pages so put it last (or explicitly route them)
```
The comment can go. In my bigger PR there are a few pages above them.
It might be clearer to split `/jwks` and `/config` as specific routes routes (and maybe even files but because they were small I combined them both in lti13Instance.ts). This was a warning/reminder for me to keep that last. Do you care how this should be resolved?

trombonekenny commented on apps/prairielearn/src/migrations/20231003211522_lti13_instance__create.sql:
```diff
@@ -0,0 +1,65 @@
+    institution_id BIGINT NOT NULL REFERENCES institutions ON UPDATE CASCADE ON DELETE SET NULL,
+    created_at timestamptz NOT NULL DEFAULT current_timestamp,
+    deleted_at timestamptz,
+    platform text,
+    name text,
+    tool_platform_name text,
+    keystore jsonb,
+    issuer_params jsonb,
+    client_params jsonb,
+    custom_params jsonb,
+    uid_attribute text,
+    uin_attribute text,
+    name_attribute text,
+    access_tokenset jsonb,
+    access_token_expires_at timestamptz
```
Maybe but at the time it's created we don't know the platform and name. I don't think it's harmful to leave most of them as `null` as different pieces set things at different times.

nwalters512 replied:
It feels a little weird to me that we can insert rows where we don't know either the platform or the name - it feels like we'd want to know what those are *before* we insert them. Maybe it'll make more sense to me when I read the code more carefully.

trombonekenny replied:
Maybe, but information like name or platform isn't required for the LTI integration to work -- and in some cases can get auto-filled after the first login. The only thing necessary is that the LTI instance exists and calling it by its ID works until we have more data later. It was a design choice to make the add simple and focus on the edit/update.

trombonekenny commented on apps/prairielearn/src/ee/pages/institutionAdminLti13/institutionAdminLti13.sql:
```diff
@@ -0,0 +1,90 @@
+SELECT
+  *
+FROM
+  pl_lti13_instances
```
Fixed in f51e1063a

trombonekenny commented on apps/prairielearn/src/ee/pages/lti13Instance/lti13Instance.ts:
```diff
@@ -0,0 +1,82 @@
+    //
+    // The file is human readable on purpose to support manual configuration in the LMS.
+
+    const lmsConfig = ltiConfig; //structuredClone(ltiConfig);
```
Looks like structuredClone was added to node in v17.0.0 and the container is still running v16.20.2

trombonekenny replied:
Switched to lodash.cloneDeep in e5bbbf9c6

trombonekenny commented on apps/prairielearn/src/ee/pages/institutionAdminLti13/institutionAdminLti13.ts:
```diff
@@ -0,0 +1,192 @@
+        institution_id: req.params.institution_id,
+        issuer_params: req.body.issuer_params,
+        platform: req.body.platform,
+        client_params,
```
Both `issuer_params` and `client_params` go directly into [openid-client's Client() and Issuer() calls](https://github.com/panva/node-openid-client/blob/main/docs/README.md#new-clientmetadata-jwks-options). My thinking was, not knowing what all of those settings might need to be for different LMS configurations, and because PL doesn't really interact with it (other than to hand it off to the library), we can store it independently and just pass it into the functions. As we learn more about what's needed for the LTI 1.3 OAuth config, we can move things more into columns: client_params is a good example of how that evolved as we're storing the `client_id` as an independent variable/column and load it into the `client_params` on save.

mwest1066 replied:
I'm fine to store these as JSON blobs for now.

trombonekenny commented on apps/prairielearn/src/ee/pages/institutionAdminLti13/institutionAdminLti13.ts:
```diff
@@ -0,0 +1,192 @@
+      });
+    }
+
+    res.redirect(req.originalUrl);
```
Fixed

trombonekenny commented:
Hi @nwalters512, I think I addressed all of the comments from the previous review. I'm available to meet on Thursday if you want to talk about any of this then.

trombonekenny commented on apps/prairielearn/src/migrations/20231003211522_lti13_instance__create.sql:
```diff
@@ -0,0 +1,27 @@
+    uin_attribute text,
+    name_attribute text,
+    access_tokenset jsonb,
+    access_token_expires_at timestamptz
```
Resolved in 39dee063b

trombonekenny commented on apps/prairielearn/assets/scripts/institutionAdminLti13.ts:
```diff
@@ -0,0 +1,27 @@
+import { onDocumentReady, decodeData } from '@prairielearn/browser-utils';
+
+onDocumentReady(() => {
+  const platformDefaults = decodeData('platform_defaults_data');
```
Addressed in 67668ddd4

nwalters512 commented on apps/prairielearn/src/ee/pages/institutionAdminLti13/institutionAdminLti13.ts:
```diff
@@ -0,0 +1,229 @@
+
+    // Handle the / (no id passed case)
+    if (typeof req.params.unsafe_lti13_instance_id === 'undefined' && lti13Instances.length > 0) {
+      return res.redirect(`lti13/${lti13Instances[0].id}`);
```
Will this do what we want? I honestly don't know what the behavior of a relative redirect is; we virtually always construct a full absolute path when redirecting.

trombonekenny replied:
It worked but easy enough to change to a absolute path in e58bb1c9

nwalters512 commented on apps/prairielearn/src/ee/pages/institutionAdminLti13/institutionAdminLti13.ts:
```diff
@@ -0,0 +1,229 @@
+      return res.redirect(`lti13/${lti13Instances[0].id}`);
+    }
+
+    const paramInstance = lti13Instances.find(
+      ({ id }) => id === req.params.unsafe_lti13_instance_id,
+    );
```
Do we want to 404 if instance isn't found?

trombonekenny replied:
Better route checking added in e58bb1c9

nwalters512 commented on apps/prairielearn/src/models/lti13Instance.ts:
```diff
@@ -0,0 +1,20 @@
+    Lti13InstanceSchema,
+  );
+
+  if (!lti13_instance) {
+    throw new Error(`LTI 1.3 instance ID ${lti13_instance_id} is unavailable`);
+  }
```
Any particular reason you're manually checking this instead of using `queryRow`? That function will automatically assert that there is a row and throw an error otherwise.

trombonekenny replied:
I believe it was so that I could customize the error message. This PR doesn't have an option to 'disable' an instance but that might have been in an earlier design and a cross over for an error that doesn't look like "invalid number of rows"

trombonekenny replied:
I just realized there's a `src/ee/models` directory so I'll move this there.

nwalters512 reviewed (approved):
Approving with a few comments!
